File: FacadeDemo/FacadeDemo.js
class Account {
  constructor(accountNumber, holdersName, balance) {
    this.accountNumber = accountNumber;
    this.holdersName = holdersName;
    this.balance = balance;
  }
}

class Bank {
  accounts = [];

  constructor(account) {
    this.accounts.push(new Account(101, "Mukesh Ambani", 100_000_000));
    this.accounts.push(new Account(102, "Gautam Adani", 200_000_000));
    this.accounts.push(new Account(103, "Anil Ambani", 100_000));
  }

  getAccount(name) {
    const searchName = name.trim().toLowerCase();
    const found = this.accounts.find(
      (account) => account.holdersName.trim().toLowerCase() === searchName
    );
    return found;
  }
}

class Rating {
  constructor(name, rating) {
    this.name = name;
    this.rating = rating;
  }
}

class CreditRating {
  ratings = [];

  constructor() {
    this.ratings.push(new Rating("Mukesh Ambani", 900));
    this.ratings.push(new Rating("Gautam Adani", 750));
    this.ratings.push(new Rating("Anil Ambani", 250));
  }

  eligibleForLoan(name) {
    const searchName = name.trim().toLowerCase();
    const person = this.ratings.find(
      (obj) => obj.name.trim().toLowerCase() === searchName
    );
    if (person) return person.rating >= 650;
    return false;
  }
}

class LoanHistory {
  constructor(name, loanAmount, paid) {
    this.name = name;
    this.loanAmount = loanAmount;
    this.paid = paid;
  }
}

class CreditHistory {
  history = [];
  constructor() {
    this.history.push(new LoanHistory("Mukesh Ambani", 100_000, "paid"));
    this.history.push(new LoanHistory("Mukesh Ambani", 500_000, "ongoing"));
    this.history.push(new LoanHistory("Gautam Ambani", 10_000_000, "ongoing"));
    this.history.push(new LoanHistory("Gautam Ambani", 15_000_000, "ongoing"));
    this.history.push(new LoanHistory("Gautam Ambani", 2_000_000, "paid"));
    this.history.push(new LoanHistory("Anil Ambani", 5_000_000, "paid"));
    this.history.push(new LoanHistory("Anil Ambani", 10_000_000, "defaulted"));
  }

  hasDefaulted(name) {
    const searchName = name.trim().toLowerCase();
    const result = this.history.find(
      (history) =>
        history.name.trim().toLowerCase() === searchName &&
        history.paid.trim().toLowerCase() === "defaulted"
    );

    return result == null;
  }
}

// const FunctionMap = {};

class LoanSystem {
  // constructor(name, amount) {
  //   this.name = name;
  //   this.amount = amount;
  // }

  applyLoan(name, amount) {
    const bank = new Bank();
    const rating = new CreditRating();
    const history = new CreditHistory();
    let approved = true;

    const conditionObj = {
      noBankAccount: {
        condition: (name) => {
          new Bank();
          !bank.getAccount(name);
        },
        errMsg: (name) => `Application ${name} does not have a bank account`,
      },
      notEligible: {
        condition: (name) => {
          const rating = new Rating();
          !rating.eligibleForLoan(name);
        },
        errorMessage: (name) =>
          `Application ${name} is not eligible for a loan`,
      },
      hasDefaulted: {
        condition: (history, name) => !history.hasDefaulted(name),
        errorMessage: (name) => `Application ${name} has defaulted before`,
      },
      insufficientFunds: {
        condition: (bank, name, amount) =>
          bank.getAccount(name).balance < amount / 10,
        errorMessage: (name) =>
          `Application ${name} does not have sufficient funds`,
      },
    };

    // if (!bank.getAccount(name)) {
    //   console.error(`Application ${name} does not have a bank account`);
    //   approved = false;
    // }

    // if (approved && !rating.eligibleForLoan(name)) {
    //   console.error(`Application ${name} is not eligible for loan`);
    //   approved = false;
    // }

    // if (approved && !history.hasDefaulted(name)) {
    //   console.error(`Application ${name} has defaulted before`);
    //   approved = false;
    // }

    // if (approved && bank.getAccount(name).balance < amount / 10) {
    //   console.error(`Application ${name} does not have sufficient funds`);
    //   approved = false;
    // }

    for (const conditionKey in conditionObj) {
      const { condition, errMsg } = conditionObj[conditionKey];
    }

    console.log(
      `Loan of amount ${amount} of ${name} has${
        approved ? " " : " not "
      }been approved`
    );
  }
}

// const name = "Anil Ambani";
// const amount = 100_000;

const loanSystem = new LoanSystem();
loanSystem.applyLoan("Mukesh Ambani", 100_000);

File: FacadeDemo/FacadeDemo.py

class Account:
    def __init__(self, account_number: int, holders_name: str, balance: float) -> None:
        self.account_number: int = account_number
        self.holders_name: str = holders_name
        self.balance: float = balance


class Bank:
    def __init__(self, account: Account | None = None) -> None:
        self.accounts: list[Account] = []
        self.accounts.append(Account(101, "Mukesh Ambani", 100_000_000))
        self.accounts.append(Account(102, "Gautam Adani", 200_000_000))
        self.accounts.append(Account(103, "Anil Ambani", 100_000))

    def get_account(self, name: str) -> Account | None:
        search_name: str = name.strip().lower()
        found = next(filter(lambda account: account.holders_name.strip().lower()
                            == search_name, self.accounts), None)
        return found


class Rating:
    def __init__(self, name: str, rating: int) -> None:
        self.name: str = name
        self.rating: int = rating


class CreditRating:
    def __init__(self) -> None:
        self.ratings: list[Rating] = []
        self.ratings.append(Rating("Mukesh Ambani", 900))
        self.ratings.append(Rating("Gautam Adani", 750))
        self.ratings.append(Rating("Anil Ambani", 250))

    def eligible_for_loan(self, name: str) -> bool:
        search_name = name.strip().lower()
        person = next(
            filter(lambda rating: rating.name.strip().lower() == search_name, self.ratings), None)
        if person is not None:
            return person.rating >= 650
        return False


class LoanHistory:
    def __init__(self, name: str, loan_amount: int, paid: str) -> None:
        self.name: str = name
        self.loan_amount: int = loan_amount
        self.paid: str = paid


class CreditHistory:
    def __init__(self) -> None:
        self.history: list[LoanHistory] = []
        self.history.append(LoanHistory("Mukesh Ambani", 100_000, "paid"))
        self.history.append(LoanHistory("Mukesh Ambani", 500_000, "ongoing"))
        self.history.append(LoanHistory(
            "Gautam Ambani", 10_000_000, "ongoing"))
        self.history.append(LoanHistory(
            "Gautam Ambani", 15_000_000, "ongoing"))
        self.history.append(LoanHistory("Gautam Ambani", 2_000_000, "paid"))
        self.history.append(LoanHistory("Anil Ambani", 5_000_000, "paid"))
        self.history.append(LoanHistory(
            "Anil Ambani", 10_000_000, "defaulted"))

    def has_defaulted(self, name: str) -> bool:
        search_name = name.strip().lower()
        result = next(filter(lambda history:
                             history.name.strip().lower() == search_name
                             and history.paid.strip().lower() == "defaulted", self.history), None)
        return result is not None


name = "Anil Ambani"
amount = 100_000
approved = True

bank = Bank()
rating = CreditRating()
history = CreditHistory()
account: Account | None = bank.get_account(name)
if not bank.get_account(name):
    print(f"Application {name} does not have a bank account")
    approved = False
if approved and not rating.eligible_for_loan(name):
    print(f"Application {name} is not eligible for loan")
    approved = False
if approved and history.has_defaulted(name):
    print(f"Application {name} has defaulted before")
    approved = False
if approved and account is not None:
    if account.balance < amount / 10:
        print(f"Application {name} does not have sufficient funds")
        approved = False

print(
    f"Loan of amount {amount} of {name} has{ ' ' if approved else ' not '}been approved")
